src/Shared/CommandLineUtils/CommandLine/ConsoleTable.cs

```diff
@@ -0,0 +1,81 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+
+namespace Microsoft.Extensions.CommandLineUtils;
+
+internal sealed class ConsoleTable
+{
+    private readonly List<string> _columns = new();
+    private readonly List<object[]> _rows = new();
+
+    public void AddColumns(params string[] names)
+    {
+        _columns.AddRange(names);
+    }
+
+    public void AddRows(params object[] values)
+    {
+        if (values == null)
+        {
+            throw new ArgumentNullException(nameof(values));
+        }
+
+        if (!_columns.Any())
+        {
+            throw new Exception("Columns must be set before rows can be added.");
+        }
+
+        if (_columns.Count != values.Length)
+        {
+            throw new Exception(
+                $"The number columns in the table '{_columns.Count}' does not match the number of columns in the row '{values.Length}'.");
+        }
+
+        _rows.Add(values);
+    }
+
+    public void Write()
+    {
+        var builder = new StringBuilder();
+
+        var maxColumnLengths = _columns
+            .Select((t, i) => _rows.Select(x => x[i])
+                .Union(new[] { _columns[i] })
+                .Where(x => x != null)
+                .Select(x => x!.ToString()!.Length).Max())
+            .ToList();
+
+        var formatRow = Enumerable.Range(0, _columns.Count)
+            .Select(i => " | {" + i + ", " + maxColumnLengths[i] + "}")
+            .Aggregate((previousRowColumn, nextRowColumn) => previousRowColumn + nextRowColumn) + " |";
+
+        var maxRowLength = Math.Max(0, _rows.Any() ? _rows.Max(row => string.Format(CultureInfo.InvariantCulture, formatRow, row).Length) : 0);
+        var columnHeaders = string.Format(CultureInfo.InvariantCulture, formatRow, _columns.ToArray());
+
+        var maxTotalLength = Math.Max(maxRowLength, columnHeaders.Length);
+
+        var formattedRows = _rows.Select(row => string.Format(CultureInfo.InvariantCulture, formatRow, row)).ToList();
```

Comment:
Might be nice to avoid `string.Format`ting the rows twice (done above for `maxRowLength`)